File: solution/1609. Even Odd Tree/1609_01.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isEvenOddTree(TreeNode* root) {
        queue<TreeNode*> q;
        q.push(root);
        int increasing = true;
        while (!q.empty()) {
            int pre = increasing ? 0 : 1e6 + 1;
            for (int i = q.size(); i > 0; --i) {
                auto node = q.front(); q.pop();
                if (increasing) {
                    if (pre >= node->val || node->val % 2 == 0)
                        return false;
                } else {
                    if (pre <= node->val || node->val % 2 == 1)
                        return false;
                }
                pre = node->val;
                if (node->left)
                    q.push(node->left);
                if (node->right)
                    q.push(node->right);
            }
            increasing = !increasing;
        }
        return true;
    }
};

File: solution/1609. Even Odd Tree/1609_01.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isEvenOddTree(self, root: Optional[TreeNode]) -> bool:
        q = [root]
        inc = True
        while q:
            pre = -1 if inc else 10**6 + 1
            for _ in range(len(q)):
                node = q.pop(0)
                if inc:
                    if pre >= node.val or node.val % 2 == 0:
                        return False
                else:
                    if pre <= node.val or node.val % 2 == 1:
                        return False
                pre = node.val
                if node.left:
                    q.append(node.left)
                if node.right:
                    q.append(node.right)
            inc = not inc
        return True
    
